src/Infrastructure.EntityFramework/Repositories/CollectionRepository.cs

```diff
@@ -233,17 +233,42 @@ public async Task ReplaceAsync(Core.Entities.Collection collection, IEnumerable<
         }
     }
 
-    public async Task UpdateUsersAsync(Guid id, IEnumerable<SelectionReadOnly> users)
+    public async Task UpdateUsersAsync(Guid id, IEnumerable<SelectionReadOnly> requestedUsers)
     {
         using (var scope = ServiceScopeFactory.CreateScope())
         {
             var dbContext = GetDatabaseContext(scope);
-            var procedure = new CollectionUserUpdateUsersQuery(id, users);
-            var updateData = await procedure.Update.BuildInMemory(dbContext);
-            dbContext.UpdateRange(updateData);
-            var insertData = await procedure.Insert.BuildInMemory(dbContext);
-            await dbContext.AddRangeAsync(insertData);
-            dbContext.RemoveRange(await procedure.Delete.Run(dbContext).ToListAsync());
+
+            var existingCollectionUsers = await dbContext.CollectionUsers
+                .Where(cu => cu.CollectionId == id)
+                .ToListAsync();
+
+            foreach (var requestedUser in requestedUsers)
+            {
+                var existingCollectionUser = existingCollectionUsers.FirstOrDefault(cu => cu.OrganizationUserId == requestedUser.Id);
+                if (existingCollectionUser == null)
+                {
+                    // This is a brand new entry
+                    dbContext.CollectionUsers.Add(new CollectionUser
+                    {
+                        CollectionId = id,
+                        OrganizationUserId = requestedUser.Id,
+                        HidePasswords = requestedUser.HidePasswords,
+                        ReadOnly = requestedUser.ReadOnly,
+                    });
+                    break;
```

Comment:
Isn't this going to break out of the loop after the first user and not assign any more after that?